File: src/lib/monitoring/health.ts
// ============================================================
// PersonaForge — Health Check Service
// ============================================================
// Aggregates health from: DB, cache, AI providers, ML engine.
// Used by /api/health endpoint and external uptime monitors.
// ============================================================

import { db } from "@/lib/db";
import { cache } from "@/lib/cache/redis";
import { getActiveProviderName } from "@/lib/ai/factory";
import { logger } from "./logger";

export interface HealthStatus {
  status: "healthy" | "degraded" | "unhealthy";
  timestamp: string;
  version: string;
  uptime: number;
  services: {
    database: ServiceHealth;
    cache: ServiceHealth;
    aiProvider: ServiceHealth;
    ml: ServiceHealth;
  };
}

interface ServiceHealth {
  status: "up" | "down" | "degraded";
  latencyMs?: number;
  details?: Record<string, unknown>;
  error?: string;
}

const startedAt = Date.now();

export async function checkHealth(): Promise<HealthStatus> {
  const services = await Promise.all([
    checkDatabase(),
    checkCache(),
    checkAIProvider(),
    checkML(),
  ]);

  const [database, cacheHealth, aiProvider, ml] = services;

  const anyDown = services.some((s) => s.status === "down");
  const anyDegraded = services.some((s) => s.status === "degraded");

  const overall: "healthy" | "degraded" | "unhealthy" = anyDown
    ? "unhealthy"
    : anyDegraded
    ? "degraded"
    : "healthy";

  return {
    status: overall,
    timestamp: new Date().toISOString(),
    version: process.env.APP_VERSION ?? "1.0.0",
    uptime: Date.now() - startedAt,
    services: { database, cache: cacheHealth, aiProvider, ml },
  };
}

async function checkDatabase(): Promise<ServiceHealth> {
  const start = Date.now();
  try {
    // Simple ping query
    await db.$queryRaw`SELECT 1`;
    return {
      status: "up",
      latencyMs: Date.now() - start,
      details: { provider: "postgres" },
    };
  } catch (err) {
    logger.error("DB health check failed", {}, err as Error);
    return {
      status: "down",
      latencyMs: Date.now() - start,
      error: (err as Error).message,
    };
  }
}

async function checkCache(): Promise<ServiceHealth> {
  const start = Date.now();
  try {
    const health = await cache.health();
    return {
      status: health.redis ? "up" : "degraded",
      latencyMs: Date.now() - start,
      details: health,
    };
  } catch (err) {
    return {
      status: "degraded",
      latencyMs: Date.now() - start,
      error: (err as Error).message,
    };
  }
}

async function checkAIProvider(): Promise<ServiceHealth> {
  try {
    const name = await getActiveProviderName();
    return {
      status: name === "template" ? "degraded" : "up",
      details: { activeProvider: name },
    };
  } catch (err) {
    return {
      status: "down",
      error: (err as Error).message,
    };
  }
}

async function checkML(): Promise<ServiceHealth> {
  // ML engine is in-process; always up if we got here
  return {
    status: "up",
    details: { modules: ["intent", "persona", "causal", "counterfactual", "bandit"] },
  };
}

File: backend/app/api/health.py
"""Health endpoint."""
from datetime import datetime, timezone

from fastapi import APIRouter, Depends
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app import __version__
from app.config import settings
from app.database.session import get_db
from app.database.redis import get_redis

router = APIRouter()


@router.get("")
async def health(db: AsyncSession = Depends(get_db)):
    """Aggregated health check."""
    services = {}

    # Database
    try:
        await db.execute(text("SELECT 1"))
        services["database"] = {"status": "up"}
    except Exception as e:
        services["database"] = {"status": "down", "error": str(e)}

    # Redis
    try:
        redis = await get_redis()
        await redis.ping()
        services["cache"] = {"status": "up"}
    except Exception as e:
        services["cache"] = {"status": "down", "error": str(e)}

    # AI provider
    if settings.anthropic_api_key:
        ai = "claude"
    elif settings.gemini_api_key:
        ai = "gemini"
    elif settings.openai_api_key:
        ai = "openai"
    else:
        ai = "template"
    services["ai_provider"] = {"status": "up" if ai != "template" else "degraded", "active": ai}

    services["ml"] = {"status": "up", "modules": ["intent", "persona", "causal", "counterfactual", "bandit"]}

    any_down = any(s["status"] == "down" for s in services.values())
    any_degraded = any(s["status"] == "degraded" for s in services.values())

    overall = "unhealthy" if any_down else ("degraded" if any_degraded else "healthy")
    return {
        "status": overall,
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "version": __version__,
        "services": services,
    }
